### PR TITLE
Remove palloc() call in ResWaitOnLock

### PR DESCRIPTION
**Pull request by dh-cloud**
When i investigated another one assertion failure in ResLockAcquire, i 
found this possible issue.

If palloc() failed in ResWaitOnLock due to OOM, it will trigger
a PANIC "Waiting on lock already held" because partitionLock is
not released correctly.

The first attempt is to add a try/catch surrounding for this
palloc() call, and release the partitionLock before rethrowing
error. But it is not good as ResIncrementRemove is not called.

So use local char new_status[160], length 160 is enough.

----

Steps to reproduce:

Add a fault at the palloc() position:
```
#ifdef FAULT_INJECTOR
    SIMPLE_FAULT_INJECTOR("reslock_acquire_before_wait");
#endif
```

```sql
-- superuser:
# CREATE USER test PASSWORD '123456';
# ALTER RESOURCE QUEUE pg_default WITH (ACTIVE_STATEMENTS=1);
# SELECT gp_inject_fault('reslock_acquire_before_wait', 'error', 1 /*master dbid*/);

-- test user session 1:
> SELECT pg_sleep(120);

-- test user session 2:
> SELECT version();
PANIC:  Waiting on lock already held! (lwlock.c:668)
server closed the connection unexpectedly
```

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by dh-cloud** [comment]
If backpatch it to 6X, 6X needs also https://github.com/greenplum-db/gpdb/pull/10585/files, because `cdbcomponent_cleanupIdleQEs` has palloc calls.

**Comment by SmartKeyerror** [comment]
@dh-cloud 

Hi~ Thanks very much for your contribution! 

I noticed that the concourse-ci had failed both at icw-planner and icw-gporca, there have some test cases failed, so you should run `make installcheck` at `/src/test/isolation2` to find out what's wrong.

I'm just posting a small part of the error message for you to help you locate the problem:

```
--- /tmp/build/d62a0504/gpdb_src/src/test/isolation2/expected/resource_queue.out	2022-06-22 02:43:48.674937536 +0000

+++ /tmp/build/d62a0504/gpdb_src/src/test/isolation2/results/resource_queue.out	2022-06-22 02:43:48.678937916 +0000

@@ -59,29 +59,30 @@

 END

 

 2<:  <... completed>

- ?column? 

-----------

- 1        

-(1 row)

+FATAL:  Unexpected internal error (assert.c:48)

+DETAIL:  BadArgument("!(((context) != ((void *)0) && (((((const Node*)((context)))->type) == T_AllocSetContext) || ((((const Node*)((context)))->type) == T_SlabContext) || ((((const Node*)((context)))->type) == T_GenerationContext))))", File: "../../../../src/include/utils/memutils.h", Line: 179)

+server closed the connection unexpectedly

+	This probably means the server terminated abnormally

+	before or while processing the request.

 2:END;

-END

+unknown encoding:

 

 3<:  <... completed>

- ?column? 

-----------

- 1        

-(1 row)

+server closed the connection unexpectedly

+	This probably means the server terminated abnormally

+	before or while processing the request.

 3:END;

-END

+unknown encoding:

 

 -- Sanity check: Ensure that the resource queue is now empty.

 0: SELECT rsqcountlimit, rsqcountvalue from pg_resqueue_status WHERE rsqname = 'rq_concurrency_test';

- rsqcountlimit | rsqcountvalue 

----------------+---------------

- 1             | 0             

-(1 row)

+server closed the connection unexpectedly

+	This probably means the server terminated abnormally

+	before or while processing the request.

```

File: src/backend/utils/resscheduler/resqueue.c [hunk 1]
```diff
@@ -1077,18 +1077,21 @@ ResWaitOnLock(LOCALLOCK *locallock, ResourceOwner owner, ResPortalIncrement *inc
...
-		new_status = (char *) palloc(len + 8 + 1);
-		memcpy(new_status, old_status, len);
-		strcpy(new_status + len, " queuing");
-		set_ps_display(new_status, false);		/* truncate off " queuing" */
+		snprintf(new_status, sizeof(new_status), "%.*s queuing",
+				 len, old_status);
+		set_ps_display(new_status, false);
+
+		/* Truncate off " queuing" */
+		Assert(len < sizeof(new_status));
```
Comment (higuoxing) [inline src/backend/utils/resscheduler/resqueue.c:1094]: Is `len` (the size of `old_status`) always less than `160` (`sizeof(new_status)`)? If not, when the length `old_status` is greater than `160`, the content of `new_status` looks like: `"very long string from old_status ..."` and the appended `"queuing"` is truncated off which makes it difficult for us to figure out the status.

What about the following change?

```c
len = min(len, 160 - 9); // reserve some space for " queuing".
snprintf(new_status, sizeof(new_status), "%.*s queuing", len, old_status);
set_ps_display(new_status, false);
new_status[len] = '\0'; // truncate off "queuing" for the next time display.
```
When the `old_status` exceeds the `new_status` buffer size, the content of `new_status` is: `"very long string from old_status ... queuing"`. We know that the status is queuing and we can get rid of the assertion `Assert(len < sizeof(new_status));`.

**Review by higuoxing — APPROVED** [review]
LGTM, thanks!

**Pull request merged**